commit ebb19ecb49daec2548f940d5c81115cd84c126cd
Author: Mario Campbell R <mariocampbellr@gmail.com>
Date:   9 weeks ago

    feat(spectre): add spectre and fix bugs

diff --git a/nvim/.config/nvim/after/plugin/gitsigns.lua b/nvim/.config/nvim/after/plugin/gitsigns.lua
index bb73f2f..53db7cf 100644
--- a/nvim/.config/nvim/after/plugin/gitsigns.lua
+++ b/nvim/.config/nvim/after/plugin/gitsigns.lua
@@ -20,9 +20,9 @@ gitsigns.setup {
     follow_files = true,
   },
   attach_to_untracked = true,
-  current_line_blame = true, -- Toggle with `:Gitsigns toggle_current_line_blame`
+  current_line_blame = false, -- Toggle with `:Gitsigns toggle_current_line_blame`
   current_line_blame_opts = {
-    virt_text = true,
+    virt_text = false,
     virt_text_pos = "eol", -- 'eol' | 'overlay' | 'right_align'
     delay = 1000,
     ignore_whitespace = false,
diff --git a/nvim/.config/nvim/after/plugin/notify.lua b/nvim/.config/nvim/after/plugin/notify.lua
index e69c157..78c6cb3 100644
--- a/nvim/.config/nvim/after/plugin/notify.lua
+++ b/nvim/.config/nvim/after/plugin/notify.lua
@@ -1,29 +1,41 @@
-require("notify").setup {
+local status_ok, notify = pcall(require, "notify")
+if not status_ok then
+  return
+end
+
+local icons = require "user.icons"
+
+notify.setup {
   -- Animation style (see below for details)
   stages = "fade_in_slide_out",
+
   -- Function called when a new window is opened, use for changing win settings/config
   on_open = nil,
+
   -- Function called when a window is closed
   on_close = nil,
+
   -- Render function for notifications. See notify-render()
   render = "default",
+
   -- Default timeout for notifications
-  timeout = 5000,
-  -- Max number of columns for messages
-  max_width = nil,
-  -- Max number of lines for a message
-  max_height = nil,
+  timeout = 100,
+
   -- For stages that change opacity this is treated as the highlight behind the window
-  -- Set this to either a highlight group, an RGB hex value e.g. "#000000" or a function returning an RGB code for dynamic values
-  background_colour = "Normal",
+  -- Set this to either a highlight group or an RGB hex value e.g. "#000000"
+  background_colour = "Pmenu",
+
   -- Minimum width for notification windows
-  minimum_width = 50,
+  minimum_width = 20,
+
   -- Icons for the different levels
   icons = {
-    ERROR = "",
-    WARN = "",
-    INFO = "",
-    DEBUG = "",
-    TRACE = "✎",
+    ERROR = icons.diagnostics.Error,
+    WARN = icons.diagnostics.Warning,
+    INFO = icons.diagnostics.Information,
+    DEBUG = icons.ui.Bug,
+    TRACE = icons.ui.Pencil,
   },
 }
+
+vim.notify = notify
diff --git a/nvim/.config/nvim/after/plugin/nvim-tree.lua b/nvim/.config/nvim/after/plugin/nvim-tree.lua
index 1ed401a..14a007d 100644
--- a/nvim/.config/nvim/after/plugin/nvim-tree.lua
+++ b/nvim/.config/nvim/after/plugin/nvim-tree.lua
@@ -13,8 +13,8 @@ vim.g.nvim_tree_icons = {
     ignored = "◌",
   },
   folder = {
-    arrow_open = "▿",
-    arrow_closed = "▹",
+    arrow_open = "▼",
+    arrow_closed = "▶︎",
     default = "",
     open = "",
     empty = "",
@@ -27,13 +27,6 @@ vim.g.nvim_tree_special_files = {
   "README.md" == 0,
 }
 
-vim.g.nvim_tree_show_icons = {
-  git = 1,
-  folders = 1,
-  files = 1,
-  folders_arrows = 0,
-}
-
 vim.g.nvim_tree_create_in_closed_folder = 1
 
 local status_ok, nvim_tree = pcall(require, "nvim-tree")
@@ -93,7 +86,7 @@ nvim_tree.setup {
     timeout = 500,
   },
   view = {
-    width = 30,
+    width = 50,
     height = 30,
     hide_root_folder = false,
     side = "left",
diff --git a/nvim/.config/nvim/after/plugin/spectre.lua b/nvim/.config/nvim/after/plugin/spectre.lua
new file mode 100644
index 0000000..50d5628
--- /dev/null
+++ b/nvim/.config/nvim/after/plugin/spectre.lua
@@ -0,0 +1,150 @@
+local status_ok, spectre = pcall(require, "spectre")
+if not status_ok then
+  return
+end
+
+spectre.setup({
+  color_devicons = true,
+  open_cmd = 'vnew',
+  live_update = false, -- auto excute search again when you write any file in vim
+  line_sep_start = '┌-----------------------------------------',
+  result_padding = '¦  ',
+  line_sep       = '└-----------------------------------------',
+  highlight = {
+      ui = "String",
+      search = "DiffChange",
+      replace = "DiffDelete"
+  },
+  mapping={
+    ['toggle_line'] = {
+        map = "dd",
+        cmd = "<cmd>lua require('spectre').toggle_line()<CR>",
+        desc = "toggle current item"
+    },
+    ['enter_file'] = {
+        map = "<cr>",
+        cmd = "<cmd>lua require('spectre.actions').select_entry()<CR>",
+        desc = "goto current file"
+    },
+    ['send_to_qf'] = {
+        map = "<leader>q",
+        cmd = "<cmd>lua require('spectre.actions').send_to_qf()<CR>",
+        desc = "send all item to quickfix"
+    },
+    ['replace_cmd'] = {
+        map = "<leader>c",
+        cmd = "<cmd>lua require('spectre.actions').replace_cmd()<CR>",
+        desc = "input replace vim command"
+    },
+    ['show_option_menu'] = {
+        map = "<leader>o",
+        cmd = "<cmd>lua require('spectre').show_options()<CR>",
+        desc = "show option"
+    },
+    ['run_replace'] = {
+        map = "<leader>R",
+        cmd = "<cmd>lua require('spectre.actions').run_replace()<CR>",
+        desc = "replace all"
+    },
+    ['change_view_mode'] = {
+        map = "<leader>v",
+        cmd = "<cmd>lua require('spectre').change_view()<CR>",
+        desc = "change result view mode"
+    },
+    ['toggle_live_update']={
+      map = "tu",
+      cmd = "<cmd>lua require('spectre').toggle_live_update()<CR>",
+      desc = "update change when vim write file."
+    },
+    ['toggle_ignore_case'] = {
+      map = "ti",
+      cmd = "<cmd>lua require('spectre').change_options('ignore-case')<CR>",
+      desc = "toggle ignore case"
+    },
+    ['toggle_ignore_hidden'] = {
+      map = "th",
+      cmd = "<cmd>lua require('spectre').change_options('hidden')<CR>",
+      desc = "toggle search hidden"
+    },
+    -- you can put your mapping here it only use normal mode
+  },
+  find_engine = {
+    -- rg is map with finder_cmd
+    ['rg'] = {
+      cmd = "rg",
+      -- default args
+      args = {
+        '--color=never',
+        '--no-heading',
+        '--with-filename',
+        '--line-number',
+        '--column',
+      } ,
+      options = {
+        ['ignore-case'] = {
+          value= "--ignore-case",
+          icon="[I]",
+          desc="ignore case"
+        },
+        ['hidden'] = {
+          value="--hidden",
+          desc="hidden file",
+          icon="[H]"
+        },
+        ['exact'] = {
+          value="--fixed-strings",
+          desc="Exact Match",
+          icon="[E]"
+        },
+        -- you can put any rg search option you want here it can toggle with
+        -- show_option function
+      }
+    },
+    ['ag'] = {
+      cmd = "ag",
+      args = {
+        '--vimgrep',
+        '-s'
+      } ,
+      options = {
+        ['ignore-case'] = {
+          value= "-i",
+          icon="[I]",
+          desc="ignore case"
+        },
+        ['hidden'] = {
+          value="--hidden",
+          desc="hidden file",
+          icon="[H]"
+        },
+      },
+    },
+  },
+  replace_engine={
+      ['sed']={
+          cmd = "sed",
+          args = nil
+      },
+      options = {
+        ['ignore-case'] = {
+          value= "--ignore-case",
+          icon="[I]",
+          desc="ignore case"
+        },
+      }
+  },
+  default = {
+      find = {
+          --pick one of item in find_engine
+          cmd = "rg",
+          options = {"ignore-case"}
+      },
+      replace={
+          --pick one of item in replace_engine
+          cmd = "sed"
+      }
+  },
+  replace_vim_cmd = "cdo",
+  is_open_target_win = true, --open file on opener window
+  is_insert_mode = false  -- start open panel on is_insert_mode
+})
diff --git a/nvim/.config/nvim/after/plugin/todo-comments.lua b/nvim/.config/nvim/after/plugin/todo-comments.lua
index c0ab5e7..ee3957e 100644
--- a/nvim/.config/nvim/after/plugin/todo-comments.lua
+++ b/nvim/.config/nvim/after/plugin/todo-comments.lua
@@ -24,7 +24,7 @@ todo_comments.setup {
     INFO = { icon = icons.diagnostics.Information, color = hint_blue },
   },
   highlight = {
-    before = "", -- "fg" or "bg" or empty
+    before = "fg", -- "fg" or "bg" or empty
     keyword = "fg", -- "fg", "bg", "wide" or empty. (wide is the same as bg, but will also highlight surrounding characters)
     after = "fg", -- "fg" or "bg" or empty
     pattern = [[.*<(KEYWORDS)\s*:]], -- pattern or table of patterns, used for highlightng (vim regex)
diff --git a/nvim/.config/nvim/init.lua b/nvim/.config/nvim/init.lua
index f286f15..e2e00e1 100644
--- a/nvim/.config/nvim/init.lua
+++ b/nvim/.config/nvim/init.lua
@@ -5,6 +5,7 @@ require "user.colorscheme"
 require "user.telescope"
 require "user.cmp"
 require "user.lsp"
+require "user.spectre"
 
 require "user.autocommands"
 require "user.ts-context"
diff --git a/nvim/.config/nvim/lua/user/cmp.lua b/nvim/.config/nvim/lua/user/cmp.lua
index 182b7ef..f44e7f4 100644
--- a/nvim/.config/nvim/lua/user/cmp.lua
+++ b/nvim/.config/nvim/lua/user/cmp.lua
@@ -102,7 +102,6 @@ cmp.setup {
     { name = "buffer", keyword_length = 5 },
     { name = "path" },
     { name = "npm", keyword_length = 4 },
-    { name = "emoji" },
   },
   confirm_opts = {
     behavior = cmp.ConfirmBehavior.Replace,
diff --git a/nvim/.config/nvim/lua/user/icons.lua b/nvim/.config/nvim/lua/user/icons.lua
index e54be29..3bab4fa 100644
--- a/nvim/.config/nvim/lua/user/icons.lua
+++ b/nvim/.config/nvim/lua/user/icons.lua
@@ -7,25 +7,20 @@ if vim.fn.has "mac" == 1 then
   return {
     kind = {
       Text = "",
-      -- Method = "m",
+      Method = "m",
       Function = "",
       Constructor = "",
-      Method = "",
       Field = "",
-      -- Variable = "",
-      Variable = "",
+      Variable = "",
       Class = "",
       Interface = "",
-      -- Module = "",
-      Module = "",
+      Module = "",
       Property = "",
       Unit = "",
       Value = "",
       Enum = "",
-      -- Keyword = "",
-      Keyword = "",
-      -- Snippet = "",
-      Snippet = "",
+      Keyword = "",
+      Snippet = "",
       Color = "",
       File = "",
       Reference = "",
@@ -87,8 +82,8 @@ if vim.fn.has "mac" == 1 then
       Note = "",
       BookMark = "",
       Pencil = "",
-      -- ChevronRight = "",
-      ChevronRight = ">",
+      ChevronRight = "",
+      ChevronDown = "",
       Table = "",
       Calendar = "",
     },
diff --git a/nvim/.config/nvim/lua/user/keymaps.lua b/nvim/.config/nvim/lua/user/keymaps.lua
index dd1543c..86ebe39 100644
--- a/nvim/.config/nvim/lua/user/keymaps.lua
+++ b/nvim/.config/nvim/lua/user/keymaps.lua
@@ -7,6 +7,7 @@ local keymap = vim.api.nvim_set_keymap
 
 --Remap space as leader key
 keymap("", "<Space>", "<Nop>", opts)
+keymap("n", "Q", "<Nop>", opts)
 vim.g.mapleader = " "
 vim.g.maplocalleader = " "
 
@@ -76,11 +77,11 @@ keymap("x", "<A-j>", ":move '>+1<CR>gv-gv", opts)
 keymap("x", "<A-k>", ":move '<-2<CR>gv-gv", opts)
 
 -- Custom
-keymap("n", "Q", "<cmd>Bdelete!<CR>", opts)
 keymap("n", "<F1>", "<cmd>LspToggleAutoFormat<CR>", { noremap = true })
 keymap("n", "<F3>", "<cmd>e .<cr>", opts)
-keymap("n", "<F4>", "<cmd>Telescope resume<cr>", opts)
-keymap("n", "<F5>", "<cmd>Telescope commands<CR>", opts)
+keymap("n", "<F4>", "<cmd>Bdelete!<CR>", opts)
+keymap("n", "<F5>", "<cmd>Telescope resume<cr>", opts)
+keymap("n", "<F6>", "<cmd>Telescope commands<CR>", opts)
 
 keymap("n", "<F7>", ":lua require('harpoon.ui').nav_file(1)<CR>", opts)
 keymap("n", "<F8>", ":lua require('harpoon.ui').nav_file(2)<CR>", opts)
diff --git a/nvim/.config/nvim/lua/user/lsp/handlers.lua b/nvim/.config/nvim/lua/user/lsp/handlers.lua
index 9d41c24..2484f6e 100644
--- a/nvim/.config/nvim/lua/user/lsp/handlers.lua
+++ b/nvim/.config/nvim/lua/user/lsp/handlers.lua
@@ -14,7 +14,7 @@ M.setup = function()
   end
 
   local config = {
-    virtual_text = false,
+    virtual_text = true,
     signs = {
       active = signs,
     },
diff --git a/nvim/.config/nvim/lua/user/plugins.lua b/nvim/.config/nvim/lua/user/plugins.lua
index 0460e68..bafef29 100644
--- a/nvim/.config/nvim/lua/user/plugins.lua
+++ b/nvim/.config/nvim/lua/user/plugins.lua
@@ -60,6 +60,8 @@ return packer.startup(function(use)
   use "JoosepAlviste/nvim-ts-context-commentstring"
   -- Icons over nvim
   use "kyazdani42/nvim-web-devicons"
+  -- search and replace
+  use "nvim-pack/nvim-spectre"
   -- Better notifications
   use "rcarriga/nvim-notify"
   -- File navigation
@@ -95,7 +97,6 @@ return packer.startup(function(use)
   use "hrsh7th/cmp-buffer" -- buffer completions
   use "hrsh7th/cmp-path" -- path completions
   use "hrsh7th/cmp-nvim-lsp"
-  use "hrsh7th/cmp-emoji"
   use "hrsh7th/cmp-nvim-lua"
   use "David-Kunz/cmp-npm"
   use "saadparwaiz1/cmp_luasnip" -- snippet completions
diff --git a/nvim/.config/nvim/lua/user/spectre.lua b/nvim/.config/nvim/lua/user/spectre.lua
new file mode 100644
index 0000000..781a2e1
--- /dev/null
+++ b/nvim/.config/nvim/lua/user/spectre.lua
@@ -0,0 +1,8 @@
+local opts = { noremap = true, silent = true }
+local keymap = vim.api.nvim_set_keymap
+
+keymap("n", "<leader>S", "<cmd>lua require('spectre').open()<CR>", opts)
+
+keymap("n", "<leader>sw", "<cmd>lua require('spectre').open_visual({select_word=true})<CR>", opts)
+keymap("n", "<leader>s", "<cmd>lua require('spectre').open_visual()<CR>", opts)
+keymap("n", "<leader>sf", "<cmd>lua require('spectre').open_file_search({select_word= true})<CR>", opts)
diff --git a/nvim/.config/nvim/lua/user/telescope.lua b/nvim/.config/nvim/lua/user/telescope.lua
index 0e883f6..3e384b5 100644
--- a/nvim/.config/nvim/lua/user/telescope.lua
+++ b/nvim/.config/nvim/lua/user/telescope.lua
@@ -5,6 +5,7 @@ local opts = { noremap = true, silent = true }
 
 k("n", "<C-f>", ":Telescope find_files<CR>", opts)
 k("n", "<C-t>", ":TodoTelescope<CR>", opts)
+k("n", "<leader>ff", ":Telescope current_buffer_fuzzy_find<CR>", opts)
 k("n", "<leader>dd", ":Telescope diagnostics<CR>", opts)
 k("n", "<leader>gd", ":Telescope git_status<CR>", opts)
 k("n", "<C-b>", ":Telescope buffers<CR>", opts)
diff --git a/zsh/.config/zsh/.p10k.zsh b/zsh/.config/zsh/.p10k.zsh
index 8c81fea..5d27750 100644
--- a/zsh/.config/zsh/.p10k.zsh
+++ b/zsh/.config/zsh/.p10k.zsh
@@ -190,7 +190,7 @@
   # Default prompt symbol.
   typeset -g POWERLEVEL9K_PROMPT_CHAR_{OK,ERROR}_VIINS_CONTENT_EXPANSION='$'
   # Prompt symbol in command vi mode.
-  typeset -g POWERLEVEL9K_PROMPT_CHAR_{OK,ERROR}_VICMD_CONTENT_EXPANSION='❮'
+  typeset -g POWERLEVEL9K_PROMPT_CHAR_{OK,ERROR}_VICMD_CONTENT_EXPANSION='VIM'
   # Prompt symbol in visual vi mode.
   typeset -g POWERLEVEL9K_PROMPT_CHAR_{OK,ERROR}_VIVIS_CONTENT_EXPANSION='V'
   # Prompt symbol in overwrite vi mode.
diff --git a/zsh/.zshrc b/zsh/.zshrc
index 3eddede..5c33f96 100644
--- a/zsh/.zshrc
+++ b/zsh/.zshrc
@@ -28,10 +28,10 @@ source "$ZDOTDIR/zsh-functions"
 unsetopt BEEP
 
 # Normal files to source
-zsh_add_file "zsh-aliases"
 zsh_add_file "zsh-exports"
-zsh_add_file "zsh-prompt"
 zsh_add_file "zsh-vim-mode"
+zsh_add_file "zsh-aliases"
+zsh_add_file "zsh-prompt"
 zsh_add_file "zsh-websearch"
 
 # Plugins
